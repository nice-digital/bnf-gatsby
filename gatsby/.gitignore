node_modules/
.cache/
public

# dotenv environment variables file
.env
.env.*
