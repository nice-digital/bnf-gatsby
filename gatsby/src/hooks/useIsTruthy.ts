import { useCallback } from "react";

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const useIsTruthy = () =>
	useCallback(
		<TMaybe>(maybeT: TMaybe | null): maybeT is TMaybe => !!maybeT,
		[]
	);
