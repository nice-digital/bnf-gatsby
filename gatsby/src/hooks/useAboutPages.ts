import { useStaticQuery, graphql } from "gatsby";

import { type MenuPageLink, type SlugAndTitle } from "@/utils";

export interface AboutPages {
	allBnfAboutSection: {
		aboutSectionPages: SlugAndTitle[];
	};
	allBnfCautionaryAndAdvisoryGuidance: {
		labelsGuidancePages: SlugAndTitle[];
	};
}

const slugToHref = ({ title, slug }: SlugAndTitle): MenuPageLink => ({
	title,
	href: `/about/${slug}/`,
});

/**
 * Gets the lists of pages for the about section by combining:
 * - about sections
 * - cautionary/advisory labels
 * - cautionary/advisory guidance page(s)
 * from the feed.
 *
 * @returns The consolidated list of about pages
 */
export const useAboutPages = (): MenuPageLink[] => {
	const { allBnfAboutSection, allBnfCautionaryAndAdvisoryGuidance } =
		useStaticQuery<AboutPages>(
			graphql`
				query AboutPages {
					allBnfAboutSection(sort: { order: ASC }) {
						aboutSectionPages: nodes {
							slug
							title
						}
					}
					allBnfCautionaryAndAdvisoryGuidance {
						labelsGuidancePages: nodes {
							title
							slug
						}
					}
				}
			`
		);

	return [
		...allBnfAboutSection.aboutSectionPages.map(slugToHref),
		{ title: "Labels", href: "/about/labels/" },
		...allBnfCautionaryAndAdvisoryGuidance.labelsGuidancePages.map(slugToHref),
	];
};
