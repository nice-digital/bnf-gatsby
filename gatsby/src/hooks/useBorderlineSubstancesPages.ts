import { useStaticQuery, graphql } from "gatsby";

import { SID, PHPID } from "@nice-digital/gatsby-source-bnf";

import { type MenuPageLink, type SlugAndTitle } from "@/utils";

export interface BorderlineSubstancesQueryResult {
	allBnfBorderlineSubstancesTaxonomyRoot: {
		pages: {
			slug: string;
			title: string;
		}[];
	};
}

const slugToHref = ({ title, slug }: SlugAndTitle): MenuPageLink => ({
	title,
	href: `/borderline-substances/${slug}/`,
});

/**
 * Gets the lists of pages for the borderline substances section
 *
 * @returns The consolidated list of borderline substances pages
 */
export const useBorderlineSubstancesPages = (): MenuPageLink[] => {
	const { allBnfBorderlineSubstancesTaxonomyRoot } =
		useStaticQuery<BorderlineSubstancesQueryResult>(
			graphql`
				{
					allBnfBorderlineSubstancesTaxonomyRoot {
						pages: nodes {
							slug
							title
						}
					}
				}
			`
		);

	return allBnfBorderlineSubstancesTaxonomyRoot.pages.map(slugToHref);
};
