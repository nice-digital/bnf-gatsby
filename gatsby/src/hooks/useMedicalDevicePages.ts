import { useStaticQuery, graphql } from "gatsby";

import { type MenuPageLink, type SlugAndTitle } from "@/utils";

export interface MedicalDevicesQueryResult {
	allBnfMedicalDevice: {
		pages: SlugAndTitle[];
	};
}

const slugToHref = ({ title, slug }: SlugAndTitle): MenuPageLink => ({
	title,
	href: `/medical-devices/${slug}/`,
});

/**
 * Gets the lists of pages for the medical devices section
 *
 * @returns The consolidated list of medical devices pages
 */
export const useMedicalDevicePages = (): MenuPageLink[] => {
	const { allBnfMedicalDevice } = useStaticQuery<MedicalDevicesQueryResult>(
		graphql`
			{
				allBnfMedicalDevice(sort: { fields: title, order: ASC }) {
					pages: nodes {
						slug
						title
					}
				}
			}
		`
	);

	return allBnfMedicalDevice.pages.map(slugToHref);
};
