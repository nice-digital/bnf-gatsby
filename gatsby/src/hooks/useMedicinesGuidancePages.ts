import { useStaticQuery, graphql } from "gatsby";

import { type MenuPageLink, type SlugAndTitle } from "@/utils";

export interface MedicinesGuidanceQueryResult {
	allBnfGuidance: {
		pages: SlugAndTitle[];
	};
}

const slugToHref = ({ title, slug }: SlugAndTitle): MenuPageLink => ({
	title,
	href: `/medicines-guidance/${slug}/`,
});

/**
 * Gets the lists of pages for the medicines guidance section
 *
 * @returns The consolidated list of medicines guidance pages
 */
export const useMedicinesGuidancePages = (): MenuPageLink[] => {
	const { allBnfGuidance } = useStaticQuery<MedicinesGuidanceQueryResult>(
		graphql`
			{
				allBnfGuidance(sort: { fields: order, order: ASC }) {
					pages: nodes {
						slug
						title
					}
				}
			}
		`
	);

	return allBnfGuidance.pages.map(slugToHref);
};
