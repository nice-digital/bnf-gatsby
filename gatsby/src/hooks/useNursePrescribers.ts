import { useStaticQuery, graphql } from "gatsby";

import { SlugAndTitle, type MenuPageLink } from "@/utils";

export interface NursePrescribers {
	allBnfNursePrescribersFormularyIntroduction: {
		introduction: SlugAndTitle[];
	};
	allBnfNursePrescribersFormularyTreatmentSummary: {
		summaries: SlugAndTitle[];
	};
}

const slugToHref = ({ title, slug }: SlugAndTitle): MenuPageLink => ({
	title,
	href: `/nurse-prescribers-formulary/${slug}/`,
});

/**
 * Gets the lists of pages for the nurse prescribers formulary by combining:
 * - treatment summaries
 * - introduction
 * from the feed.
 *
 * @returns both the consolidated list for the details page menu, and two separate lists for the index page.
 */
export const useNursePrescribers = (): {
	menuList: MenuPageLink[];
	aboutList: MenuPageLink[];
	treatmentSummariesList: MenuPageLink[];
} => {
	const {
		allBnfNursePrescribersFormularyIntroduction,
		allBnfNursePrescribersFormularyTreatmentSummary,
	} = useStaticQuery<NursePrescribers>(
		graphql`
			query NursePrescribers {
				allBnfNursePrescribersFormularyIntroduction {
					introduction: nodes {
						slug
						title
					}
				}
				allBnfNursePrescribersFormularyTreatmentSummary {
					summaries: nodes {
						slug
						title
					}
				}
			}
		`
	);

	const menuList = [
		...allBnfNursePrescribersFormularyTreatmentSummary.summaries.map(
			slugToHref
		),
		...allBnfNursePrescribersFormularyIntroduction.introduction.map(slugToHref),
	];

	const aboutList = [
		...allBnfNursePrescribersFormularyTreatmentSummary.summaries
			.filter((summary) => summary.slug == "general-guidance")
			.map(slugToHref),
		...allBnfNursePrescribersFormularyIntroduction.introduction.map(slugToHref),
	];

	const treatmentSummariesList = [
		...allBnfNursePrescribersFormularyTreatmentSummary.summaries.map(
			slugToHref
		),
	];

	return { menuList, aboutList, treatmentSummariesList };
};
