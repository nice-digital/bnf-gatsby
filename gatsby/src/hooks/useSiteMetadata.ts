import { useStaticQuery, graphql } from "gatsby";

interface SiteMetaData {
	/** Whether the current site is BNF or BNFC */
	isBNF: boolean;
	/** Absolute URL to the site homepage, without trailing slash */
	siteUrl: string;
	/** Short site title e.g. "BNF" or "BNFC" */
	siteTitleShort: string;
	/** Full site title e.g. "British National Formulary" or "British National Formulary for Children" */
	siteTitleLong: string;
}

interface Site {
	site: {
		siteMetadata: SiteMetaData;
	};
}

/**
 * Gets the site metadata, as defined in gatsby-config
 *
 * @returns The site metadata
 */
export const useSiteMetadata = (): SiteMetaData => {
	const { site } = useStaticQuery<Site>(
		graphql`
			query SiteMetaData {
				site {
					siteMetadata {
						isBNF
						siteUrl
						siteTitleShort
						siteTitleLong
					}
				}
			}
		`
	);
	return site.siteMetadata;
};
