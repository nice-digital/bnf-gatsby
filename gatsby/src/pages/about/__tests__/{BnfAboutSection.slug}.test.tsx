import { useLocation } from "@reach/router";
import { render, screen, waitFor } from "@testing-library/react";
import { useStaticQuery } from "gatsby";

import { mockAboutPagesQueryData } from "@/hooks/useAboutPages.test";
import { useSiteMetadata } from "@/hooks/useSiteMetadata";

import AboutSectionPage, {
	type AboutSectionPageProps,
	query,
} from "../{BnfAboutSection.slug}";

const pageProps: AboutSectionPageProps = {
	data: {
		currentAboutPage: {
			title: "Changes",
			sections: [
				{
					order: 0,
					content:
						"<p>Monthly updates are provided online via Medicines Complete and the NHS Evidence portal. The changes listed below are cumulative (from one print edition to the next).</p>",
					slug: "introduction",
					title: "Introduction",
				},
				{
					order: 1,
					content:
						"<p>Significant changes made since the release of data for the print edition of BNF 82 (September 2021 â€” March 2022):</p>",
					slug: "significant-changes",
					title: "Significant changes",
				},
			],
		},
	},
	location: {
		pathname: "/about/changes/",
	},
};

describe("AboutSectionPage", () => {
	beforeEach(() => {
		(useLocation as jest.Mock).mockImplementation(
			() =>
				new URL(
					"https://bnf-gatsby-tests.nice.org.uk" + pageProps.location.pathname
				)
		);

		(useStaticQuery as jest.Mock).mockReturnValue(mockAboutPagesQueryData);
	});

	it("should match snapshot for graphql query", () => {
		expect(query).toMatchSnapshot();
	});

	it("should match snapshot for page contents", () => {
		render(<AboutSectionPage {...pageProps} />);

		expect(screen.getByRole("main")).toMatchSnapshot();
	});

	it("should set page title", async () => {
		render(<AboutSectionPage {...pageProps} />);

		await waitFor(() => {
			expect(document.title).toStartWith("Changes | About | ");
		});
	});

	it("should set meta description for known path on BNF", async () => {
		render(<AboutSectionPage {...pageProps} />);

		await waitFor(() => {
			expect(
				// eslint-disable-next-line testing-library/no-node-access
				document.querySelector(`meta[name="description"]`)
			).toHaveAttribute(
				"content",
				"Keep up to date with the latest significant changes in the BNF that are relevant to your clinical practice. Updated monthly."
			);
		});
	});

	it("should set meta description for known path on BNFC", async () => {
		(useSiteMetadata as jest.Mock).mockImplementationOnce(() => ({
			siteTitleShort: "BNFC",
		}));

		render(<AboutSectionPage {...pageProps} />);

		await waitFor(() => {
			expect(
				// eslint-disable-next-line testing-library/no-node-access
				document.querySelector(`meta[name="description"]`)
			).toHaveAttribute(
				"content",
				"Keep up to date with the latest significant changes in the BNFC that are relevant to your clinical practice. Updated monthly."
			);
		});
	});

	it("should throw error for unknown path", () => {
		expect(() => {
			render(
				<AboutSectionPage
					{...pageProps}
					location={{ pathname: "/about/unknown" }}
				/>
			);
		}).toThrowError(
			"Couldn't find meta description for page with path /about/unknown"
		);
	});
});
