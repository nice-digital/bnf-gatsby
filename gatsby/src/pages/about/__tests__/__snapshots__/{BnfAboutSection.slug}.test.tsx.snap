// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AboutSectionPage should match snapshot for graphql query 1`] = `
"
	query ($id: String) {
		currentAboutPage: bnfAboutSection(id: { eq: $id }) {
			slug
			title
			sections {
				...RecordSection
			}
		}
	}
"
`;

exports[`AboutSectionPage should match snapshot for page contents 1`] = `
<main>
  <div
    class="container"
  >
    <nav
      aria-label="Breadcrumbs"
      role="navigation"
    >
      <ol
        class="breadcrumbs"
      >
        <li
          class="breadcrumbs__crumb"
        >
          <a
            href="https://www.nice.org.uk/"
          >
            NICE
          </a>
        </li>
        <li
          class="breadcrumbs__crumb"
        >
          <a
            href="/"
          >
            BNF
          </a>
        </li>
        <li
          class="breadcrumbs__crumb"
        >
          <a
            href="/about/"
          >
            About
          </a>
        </li>
        <li
          class="breadcrumbs__crumb"
        >
          <span>
            Changes
          </span>
        </li>
      </ol>
    </nav>
    <script
      type="application/ld+json"
    >
      {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "https://www.nice.org.uk/",
        "name": "NICE"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/",
        "name": "BNF"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "@id": "/about/",
        "name": "About"
      }
    },
    {
      "@type": "ListItem",
      "position": 4,
      "item": {
        "name": "Changes"
      }
    }
  ]
}
    </script>
    <div
      class="page-header"
      id="content-start"
    >
      <h1
        class="page-header__heading"
      >
        <span>
          Changes
        </span>
      </h1>
    </div>
    <div
      class="grid grid--loose"
      data-testid="body"
    >
      <div
        class="hide-print"
        data-g="12 md:4 lg:3"
      >
        <button
          aria-expanded="false"
          aria-label="Expand menu for About"
          class="toggleButton"
          type="button"
        >
          About
           
          <svg
            aria-hidden="true"
            class="icon"
            fill="none"
            focusable="false"
            height="1em"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M478.152 174c0 10.091-3.525 18.661-10.576 25.712l-186 186c-7.243 7.243-15.909 10.864-26 10.864-10.283 0-18.853-3.621-25.712-10.864l-186-186C36.621 192.853 33 184.283 33 174c0-10.091 3.621-18.757 10.864-26L65 126.576C72.424 119.525 81.09 116 91 116c10.091 0 18.661 3.525 25.712 10.576L255.576 265.44 394.44 126.576c7.051-7.051 15.621-10.576 25.712-10.576 9.909 0 18.576 3.525 26 10.576L467.576 148c7.051 7.424 10.576 16.091 10.576 26z"
              fill="currentColor"
            />
          </svg>
        </button>
        <nav
          aria-label="About section pages"
          class="stacked-nav"
          id="collapsible-menu"
        >
          <p
            class="stacked-nav__root"
          >
            <a
              aria-current="false"
              href="/about/"
            >
              About
            </a>
          </p>
          <ul
            class="stacked-nav__list"
          >
            <li
              class="stacked-nav__list-item"
            >
              <a
                aria-current="true"
                href="/about/changes/"
              >
                <span
                  class="stacked-nav__content-wrapper"
                >
                  <span>
                    Changes
                  </span>
                </span>
              </a>
            </li>
            <li
              class="stacked-nav__list-item"
            >
              <a
                aria-current="false"
                href="/about/publication-information/"
              >
                <span
                  class="stacked-nav__content-wrapper"
                >
                  <span>
                    Publication 
                    <i>
                      information
                    </i>
                  </span>
                </span>
              </a>
            </li>
            <li
              class="stacked-nav__list-item"
            >
              <a
                aria-current="false"
                href="/about/labels/"
              >
                <span
                  class="stacked-nav__content-wrapper"
                >
                  <span>
                    Labels
                  </span>
                </span>
              </a>
            </li>
            <li
              class="stacked-nav__list-item"
            >
              <a
                aria-current="false"
                href="/about/guidance-for-cautionary-and-advisory-labels/"
              >
                <span
                  class="stacked-nav__content-wrapper"
                >
                  <span>
                    Guidance for cautionary and advisory labels
                  </span>
                </span>
              </a>
            </li>
          </ul>
        </nav>
      </div>
      <div
        data-g="12 md:8 lg:9"
      >
        <div
          class="grid grid--rev grid--loose"
        >
          <div
            data-g="12 lg:3"
          >
            <nav
              aria-labelledby="on-this-page"
              class="wrapper"
            >
              <h2
                class="heading"
                id="on-this-page"
              >
                On this page
              </h2>
              <ol
                aria-label="Jump links to sections on this page"
                class="list"
              >
                <li>
                  <a
                    href="#introduction"
                  >
                    Introduction
                  </a>
                </li>
                <li>
                  <a
                    href="#significant-changes"
                  >
                    Significant changes
                  </a>
                </li>
              </ol>
            </nav>
          </div>
          <div
            class="body"
            data-g="12 lg:9"
          >
            <section
              aria-labelledby="introduction"
            >
              <h2
                class="firstHeading"
                id="introduction"
              >
                Introduction
              </h2>
              <div>
                <p>
                  Monthly updates are provided online via Medicines Complete and the NHS Evidence portal. The changes listed below are cumulative (from one print edition to the next).
                </p>
              </div>
            </section>
            <section
              aria-labelledby="significant-changes"
            >
              <h2
                id="significant-changes"
              >
                Significant changes
              </h2>
              <div>
                <p>
                  Significant changes made since the release of data for the print edition of BNF 82 (September 2021 â€” March 2022):
                </p>
              </div>
            </section>
          </div>
        </div>
      </div>
    </div>
  </div>
</main>
`;
